### ISSUE TITLE
Problems with inline/indented comments in sections

### ISSUE THREAD

**Issue by abravalheri** [pyscaffold/configupdater#25]
Hello, I am just using this issue to document some limitations of `ConfigUpdater`.
Maybe if they cannot be fixed, it would be nice to add them to the docs?

#### No support for indented comment on beginning of a section
```python
>>> from configupdater import ConfigUpdater
>>> cu = ConfigUpdater()

>>> sample1 = """
... [section1]
...     # indented comment
... a = 42
... """

>>> cu.load_string(sample1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/abravalheri/projects/pyscaffold/.venv/lib/python3.8/site-packages/configupdater/configupdater.py", line 732, in read_string
    self.read_file(sfile, source)
  File "/home/abravalheri/projects/pyscaffold/.venv/lib/python3.8/site-packages/configupdater/configupdater.py", line 722, in read_file
    self._read(f, source)
  File "/home/abravalheri/projects/pyscaffold/.venv/lib/python3.8/site-packages/configupdater/configupdater.py", line 912, in _read
    raise e
configparser.ParsingError: Source contains parsing errors: '<string>'
        [line  3]: '   # indented comment\n'
```

Here `ConfigParser` seems to handle the case fine.
```
>>> from configparser import ConfigParser
>>> cp = ConfigParser() 
>>> cp.load_string(sample1)
>>> cp['section1']['a']
42
```

#### Inconsistent behaviour on inline comments for section titles
```python
>>> sample2 = """
... [section2]  # inline comment
... a = 42
... """

>>> cu.load_string(sample2)
>>> print(cu)
[section2]  # inline comment
a = 42

>>> cu['section2'].name = 'different'
>>> print(cu)
[different]
a = 42

```
Here the comment is gone once we change the title of the section.

## Versions and main components

* ConfigUpdater Version: 2.0
* Python Version: 3.8
* How did you install ConfigUpdater: pip

**Comment by FlorianWilhelm** [comment]
Thanks, @abravalheri, how do you always find these bugs? ;-)
I hope I fixed the indented comment problem in PR #26. 

The latter one is really tricky and I think it cannot be really fixed due to the simple concept of blocks that configupdater follows. Inline comments are not really covered by this simple concept. One would have to rewrite configupdater completely to maybe have a concept that a single line can be composed of two blocks, e.g. an option value + a comment. Not sure it's worth it though and it's gonna be a lot of work as parsers and ASTs are tricky as I had to learn the hard way.

**Comment by abravalheri** [comment]
Hi @FlorianWilhelm, thank you very much for having a look into this! 

> how do you always find these bugs? ;-)
😝 I have been playing around during the weekend with the idea of a converter from `setup.cfg` to `pyproject.toml` for https://github.com/pyscaffold/pyscaffold/discussions/336.

Regarding of second case, I understand it is weird/very "edgey", but I guess it is worthy documenting it as a limitation at least...

I think this behaviour is slightly different than when changing a option's value, because we are changing a section name, not its values/contents. If we think about options the equivalent would be:
```python
>>> sample3 = """
... [section3] 
... a = 42  # inline comment
... """

>>> cu.load_string(sample3)
>>> cu['section3']['a'].key= 'b'
>>> print(cu)
[section3]
b = 42  # inline comment
```
which keeps the inline comment, right?

Just for the sake of brainstorming, if we consider tackling this issue (we can simply accept it as a limitation), I think there might be an easier workaround:
It seems that `section._lines` actually keeps a copy of the original section header line, so something like the following could be possible:
```python
@property
def inline_header_comment(self):
     return ''.join(self._lines).split(']')[-1]
     # or _lines[0], not sure if len(self._lines) is always 1
```
then in the `def __str__` function we could do:
```diff
- s = "[{}]\n".format(self._name)
+ s = "[{}]{}\n".format(self._name, self.inline_header_comment)
```

**Comment by FlorianWilhelm** [comment]
That's a good direction. To make it more consistent with how the rest of configupdater works I guess it would be better to parse it only once and save the comment in some additional attribute. An additional property like `comment` that works like `name` could then allow to read and write the inline comment. 

 I'll hope to have a look into it this weekend as I am off now for two days of hiking. Sorry, that I haven't worked on your other mypy PR so far. It's still on my todo list :-)

**Comment by FlorianWilhelm** [comment]
@abravalheri, is it actually possible to move all the config from `setup.cfg` into `pyproject.toml`? It looks like setuptools says one still needs setup.cfg: https://setuptools.readthedocs.io/en/latest/build_meta.html

**Comment by abravalheri** [comment]
Hi @FlorianWilhelm , indeed it is not possible yet, I was just messing around.

According to the discussions in the setuptools repo the plan is precisely to use a tool/library to convert between the two, but I haven't seen one being developed yet... So I was just doing some experiments.

**Comment by FlorianWilhelm** [comment]
All parts of this issue should be fixed now by PR #45 and PR #26.
